text/0149-exponential-histogram.md

```diff
@@ -0,0 +1,162 @@
+
+```
+scale   #subBuckets base         relative error
+-5       32         1.074607828  3.66%
+-6       64         1.036632928  1.82%
+-7      128         1.018151722  0.90%
+```
+
+Let's further consider the following use cases of histograms to compare base2 and base10:
+
+1. Displaying histogram charts. With a typical base around 1.04 (around 2% relative error), there will be hundreds of buckets for a typical range over 100x. This many points can produce a reasonably smooth curve, regardless of the raw data being base2 or base10.
+2. Calculating percentiles or quantiles. This is often used in SLO monitoring. Example SLO: "99% percentile of response time need to be no more than 100ms". To minimize relative error, percentile calculation usually returns log scale mid point of a bucket. So returned percentile values won't be on 10, 100, 1000, etc., even if the histogram is base10.
+3. Answering question like "what percentage of values fall below 100". When the threshold is on 10, 100, 1000, etc, base10 histograms do give exact answer. But even in the decimal world, power of 10 numbers is a small population. For thresholds like 200, 500, etc. base10 histograms have no advantage over base2. If an exact answer is required in these cases, a user should create explicit buckets at these bounds, instead of using exponential buckets.
+
+So the "human friendliness" of base10 exponential histograms is largely an illusion. To some extent, the base2 vs. base10 question has been answered long ago: computers convert input from base10 to base2, do processing in base2, then convert final output from base2 to base10. In the histogram case, we take input in "double", which is already in [binary float point format](https://en.wikipedia.org/wiki/Double-precision_floating-point_format). Bucketing these numbers in base10 bounds effectively switches base during processing. It just adds complexity and computational cost.
+
+### Protocol support for universally mergeable histograms
+
+Now the question is if and when we add support for universally mergeable histograms. There are some options:
+
+1. No special support in protocol. Receiver derives baseScale and referenceBase if the base is close enough to a base on a referenceBase series. Implementations will decide how close is "close enough".
+2. Allow the protocol to explicitly state baseScale, with referenceBase hardwired at 2.
+3. Allow the protocol to explicitly state baseScale and an arbitrary referenceBase.
+
+This proposal is currently written with option 1, with a path to extend to option 2 later:
+
+```
+// Current. Option 1
+double base = 1;
+
+// Future. Option 2. Changing a single value into a member of a new oneof is safe and binary compatible
+oneof base_spec {
+   double base = 1;
+   sint32 base_scale = 99;  // base = 2 ^ (2^base_scale). base_scale may be negative.
+}
+```
+
+**Or should we just do option 2 right now?**
```

Comment:
+1

Leaving `referenceBase` configurable opens up a lot of possibilities to end-up with non-mergable histograms, by using different base values. You already explained how the choice of base-2 or base-10 is largely inconsequential. So if there is no clear advantage of leaving this configurable, I would opt for hard coding this value. Since floats are alreday base-2, `referenceBase=2` seems like the best option.

Reply:
This has been accepted.